src/reader/code_93_reader.ts

```diff
@@ -0,0 +1,238 @@
+                        }
+                        break;
+                    case 'd':
+                        if (nextChar >= 'A' && nextChar <= 'Z') {
+                            decodedChar = String.fromCharCode(nextCharCode + 32);
+                        } else {
+                            return null;
+                        }
+                        break;
+                    default:
+                        console.warn('* code_93_reader _decodeExtended hit default case, this may be an error', decodedChar);
+                        return null;
+                }
+                result.push(decodedChar);
+            } else {
+                result.push(char);
+            }
+        }
+        return result;
+    };
+
+    _matchCheckChar(charArray: Array<string>, index: number, maxWeight: number) {
+        const arrayToCheck = charArray.slice(0, index);
+        const length = arrayToCheck.length;
+        const weightedSums = arrayToCheck.reduce((sum, char, i) => {
+            const weight = (((i * -1) + (length - 1)) % maxWeight) + 1;
+            const value = ALPHABET.indexOf(char.charCodeAt(0));
+            return sum + (weight * value);
+        }, 0);
+
+        const checkChar = ALPHABET[(weightedSums % 47)];
+        return checkChar === charArray[index].charCodeAt(0);
+    };
+
+    _verifyChecksums(charArray: Array<string>) {
+        return this._matchCheckChar(charArray, charArray.length - 2, 20)
+            && this._matchCheckChar(charArray, charArray.length - 1, 15);
+    };
+
+    _decode(row?: Array<number>, start?: BarcodePosition | null): Barcode | null {
+        start = this._findStart();
+        if (!start) {
+            return null;
+        }
+
+        let counters = new Uint16Array([0, 0, 0, 0, 0, 0]);
+        let result: Array<string> | null = [];
+        let nextStart = this._nextSet(this._row, start.end);
+        let lastStart;
+        let decodedChar: string | null;
+        do {
+            counters = this._toCounters(nextStart, counters) as Uint16Array;
+            const pattern = this._toPattern(counters);
+            if (pattern < 0) {
+                return null;
+            }
+            decodedChar = this._patternToChar(pattern);
+            if (decodedChar === null) {
+                return null;
+            }
+            result.push(decodedChar);
+            lastStart = nextStart;
+            nextStart += ArrayHelper.sum(counters);
+            nextStart = this._nextSet(this._row, nextStart);
+        } while (decodedChar !== '*');
+        result.pop();
+
+        if (!result.length) {
+            return null;
+        }
+
+        if (!this._verifyEnd(lastStart, nextStart)) {
+            return null;
+        }
+
+        if (!this._verifyChecksums(result)) {
+            return null;
+        }
+
+        result = result.slice(0, result.length - 2);
+        // TODO: Eric really hates assigning inside an if ugliness, but it's pretty elegant here.
```

Comment:
Eric should either stick with his gut feelings or remove the TODO 😀

Reply:
agree, i intend to do a full pass on any TODOs that I leave once they are completed..